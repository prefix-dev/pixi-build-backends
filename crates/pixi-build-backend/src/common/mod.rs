//! Common utilities that are shared between the different build backends.
mod configuration;
mod requirements;
mod variants;

pub use configuration::{BuildConfigurationParams, build_configuration};
pub use requirements::{PackageRequirements, SourceRequirements, requirements};
pub use variants::compute_variants;
