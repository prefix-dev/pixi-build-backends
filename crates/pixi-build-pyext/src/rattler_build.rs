use std::{
    ffi::OsStr, io::Write, path::{Path, PathBuf}
};

use miette::IntoDiagnostic;
use pixi_build_backend::source::Source;
use pyo3::{types::{PyAnyMethods as _, PyModule}, PyResult, Python};
use rattler_build::console_utils::LoggingOutputHandler;
use tempfile::NamedTempFile;

use crate::config::RattlerBuildBackendConfig;

pub struct RattlerBuildBackend {
    pub(crate) logging_output_handler: LoggingOutputHandler,
    /// In case of rattler-build, manifest is the raw recipe
    /// We need to apply later the selectors to get the final recipe
    pub(crate) recipe_source: Source,
    pub(crate) cache_dir: Option<PathBuf>,
    pub(crate) config: RattlerBuildBackendConfig,
}

impl RattlerBuildBackend {
    /// Returns a new instance of [`RattlerBuildBackend`] by reading the
    /// manifest at the given path.
    pub fn new(
        manifest_path: &Path,
        logging_output_handler: LoggingOutputHandler,
        cache_dir: Option<PathBuf>,
        config: RattlerBuildBackendConfig,
    ) -> miette::Result<Self> {
        // Locate the recipe
        // Create a temporary file to hold the generated recipe
        // Try to place it relative to the manifest path for context, otherwise use system temp
        let mut temp_file = NamedTempFile::with_suffix(".yaml")
            .into_diagnostic()
            .map_err(|e| miette::miette!("Failed to create temporary file: {}", e))?;
        // Call the Python function
        let generated_recipe_content = Python::with_gil(|py| -> PyResult<String> {
            let module = PyModule::import(py, "recipe_generator")?;
            let func = module.getattr("generate_recipe")?;
            let result = func.call0()?;
            result.extract::<String>()
        })
        .map_err(|e| {
            miette::miette!(
                "Python error generating recipe via '{}.{}': {}",
                "recipe_generator",
                "generate_recipe",
                e
            )
        })?;

        // Write the generated recipe to the temporary file
        std::fs::write(&temp_file, &generated_recipe_content)
            .into_diagnostic()
            .map_err(|e| miette::miette!("Failed to write to temporary file: {}", e))?;

        // Load the manifest from the source directory
        let manifest_root = manifest_path.parent().expect("manifest must have a root");
        let recipe_source =
            Source::from_rooted_path(manifest_root, temp_file.path().to_path_buf()).into_diagnostic()?;

        Ok(Self {
            recipe_source,
            logging_output_handler,
            cache_dir,
            config,
        })
    }
}
