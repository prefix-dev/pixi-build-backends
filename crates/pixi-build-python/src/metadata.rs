use std::{collections::BTreeSet, path::PathBuf, str::FromStr};

use miette::Diagnostic;
use once_cell::unsync::OnceCell;
use pixi_build_backend::generated_recipe::MetadataProvider;
use pyproject_toml::PyProjectToml;
use rattler_conda_types::{ParseVersionError, Version};

#[derive(Debug, thiserror::Error, Diagnostic)]
pub enum MetadataError {
    #[error("failed to parse pyproject.toml, {0}")]
    PyProjectToml(#[from] toml_edit::de::Error),
    #[error("failed to parse version from pyproject.toml, {0}")]
    ParseVersion(ParseVersionError),
    #[error(transparent)]
    Io(#[from] std::io::Error),
}

/// An implementation of [`MetadataProvider`] that reads metadata from a
/// pyproject.toml file.
pub struct PyprojectMetadataProvider {
    manifest_root: PathBuf,
    pyproject_manifest: OnceCell<PyProjectToml>,
    ignore_pyproject_manifest: bool,
}

impl PyprojectMetadataProvider {
    /// Constructs a new `PyprojectMetadataProvider` with the given manifest root.
    ///
    /// # Arguments
    ///
    /// * `manifest_root` - The directory that contains the `pyproject.toml` file
    /// * `ignore_pyproject_manifest` - If `true`, all metadata methods will return
    ///   `None`, effectively disabling pyproject.toml metadata extraction
    pub fn new(manifest_root: impl Into<PathBuf>, ignore_pyproject_manifest: bool) -> Self {
        Self {
            manifest_root: manifest_root.into(),
            pyproject_manifest: OnceCell::default(),
            ignore_pyproject_manifest,
        }
    }

    /// Ensures that the manifest is loaded and returns the project metadata.
    fn ensure_manifest_project(&self) -> Result<Option<&pyproject_toml::Project>, MetadataError> {
        Ok(self.ensure_manifest()?.project.as_ref())
    }

    /// Ensures that the manifest is loaded
    fn ensure_manifest(&self) -> Result<&PyProjectToml, MetadataError> {
        self.pyproject_manifest.get_or_try_init(move || {
            let pyproject_toml_content =
                fs_err::read_to_string(self.manifest_root.join("pyproject.toml"))?;
            toml_edit::de::from_str(&pyproject_toml_content).map_err(MetadataError::PyProjectToml)
        })
    }

    /// Returns the set of globs that match files that influence the metadata of
    /// this package.
    ///
    /// This includes the package's own `pyproject.toml` file. These globs
    /// can be used for incremental builds to determine when metadata might
    /// have changed.
    ///
    /// # Returns
    ///
    /// A `BTreeSet` of glob patterns as strings. Common patterns include:
    /// - `"pyproject.toml"` - The package's manifest file
    pub fn input_globs(&self) -> BTreeSet<String> {
        let mut input_globs = BTreeSet::new();

        let Some(_) = self.pyproject_manifest.get() else {
            return input_globs;
        };

        // Add the pyproject.toml manifest file itself.
        input_globs.insert(String::from("pyproject.toml"));

        input_globs
    }
}

impl MetadataProvider for PyprojectMetadataProvider {
    type Error = MetadataError;

    /// Returns the package name from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the name from the project section of the pyproject.toml file.
    fn name(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .map(|proj| proj.name.clone()))
    }

    /// Returns the package version from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the version from the project section. The version string is parsed into a
    /// `rattler_conda_types::Version`.
    fn version(&self) -> Result<Option<Version>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        let Some(project) = self.ensure_manifest_project()? else {
            return Ok(None);
        };
        let Some(version) = &project.version else {
            return Ok(None);
        };
        Ok(Some(
            Version::from_str(&version.to_string()).map_err(MetadataError::ParseVersion)?,
        ))
    }

    /// Returns the package description from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the description from the project section.
    fn description(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .and_then(|proj| proj.description.clone()))
    }

    /// Returns the package homepage URL from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the homepage from the project.urls section.
    fn homepage(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .and_then(|proj| proj.urls.as_ref())
            .and_then(|urls| urls.get("Homepage").cloned()))
    }

    /// Returns the package license from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the license from the project section.
    fn license(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .and_then(|proj| proj.license.as_ref())
            .map(|license| match license {
                pyproject_toml::License::Text { text } => text.clone(),
                pyproject_toml::License::File { file } => file.to_string_lossy().to_string(),
                pyproject_toml::License::Spdx(spdx) => spdx.clone(),
            }))
    }

    /// Returns the package license file path from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the license file path from the project section if the license is specified
    /// as a file reference.
    fn license_file(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .and_then(|proj| proj.license.as_ref())
            .and_then(|license| match license {
                pyproject_toml::License::File { file } => Some(file.to_string_lossy().to_string()),
                pyproject_toml::License::Text { text: _ } => None,
                pyproject_toml::License::Spdx(_) => None,
            }))
    }

    /// Returns the package summary from the pyproject.toml manifest.
    ///
    /// This returns the same as description since pyproject.toml doesn't have
    /// a separate summary field.
    fn summary(&self) -> Result<Option<String>, Self::Error> {
        self.description()
    }

    /// Returns the package documentation URL from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the documentation URL from the project.urls section.
    fn documentation(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .and_then(|proj| proj.urls.as_ref())
            .and_then(|urls| {
                urls.get("Documentation")
                    .or_else(|| urls.get("Docs"))
                    .cloned()
            }))
    }

    /// Returns the package repository URL from the pyproject.toml manifest.
    ///
    /// If `ignore_pyproject_manifest` is true, returns `None`. Otherwise, extracts
    /// the repository URL from the project.urls section.
    fn repository(&self) -> Result<Option<String>, Self::Error> {
        if self.ignore_pyproject_manifest {
            return Ok(None);
        }
        Ok(self
            .ensure_manifest_project()?
            .and_then(|proj| proj.urls.as_ref())
            .and_then(|urls| {
                urls.get("Repository")
                    .or_else(|| urls.get("Source"))
                    .or_else(|| urls.get("Source Code"))
                    .cloned()
            }))
    }
}

#[cfg(test)]
mod tests {
    use std::{collections::HashSet, fs};

    use pixi_build_backend::generated_recipe::{GenerateRecipe, MetadataProvider};
    use rattler_conda_types::Platform;
    use tempfile::TempDir;

    use crate::{PythonGenerator, config::PythonBackendConfig, project_fixture};
    use pixi_build_types::ProjectModelV1;

    use super::*;

    /// Helper function to create a temporary directory with a pyproject.toml file
    fn create_temp_pyproject_project(pyproject_toml_content: &str) -> TempDir {
        let temp_dir = TempDir::new().expect("Failed to create temp directory");
        let pyproject_toml_path = temp_dir.path().join("pyproject.toml");
        fs::write(pyproject_toml_path, pyproject_toml_content)
            .expect("Failed to write pyproject.toml");
        temp_dir
    }

    /// Helper function to create a PyprojectMetadataProvider for testing
    fn create_metadata_provider(manifest_root: &std::path::Path) -> PyprojectMetadataProvider {
        PyprojectMetadataProvider::new(manifest_root, false)
    }

    #[test]
    fn test_basic_metadata_extraction() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "1.0.0"
description = "A test package"
license = {text = "MIT"}

[project.urls]
Homepage = "https://example.com"
Repository = "https://github.com/example/test-package"
Documentation = "https://docs.example.com"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        assert_eq!(provider.name().unwrap(), Some("test-package".to_string()));
        assert_eq!(provider.version().unwrap().unwrap().to_string(), "1.0.0");
        assert_eq!(
            provider.description().unwrap(),
            Some("A test package".to_string())
        );
        assert_eq!(provider.license().unwrap(), Some("MIT".to_string()));
        assert_eq!(
            provider.homepage().unwrap(),
            Some("https://example.com".to_string())
        );
        assert_eq!(
            provider.repository().unwrap(),
            Some("https://github.com/example/test-package".to_string())
        );
        assert_eq!(
            provider.documentation().unwrap(),
            Some("https://docs.example.com".to_string())
        );
    }

    #[test]
    fn test_license_from_file() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "1.0.0"
license = {file = "LICENSE.txt"}
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        assert_eq!(provider.license().unwrap(), Some("LICENSE.txt".to_string()));
        assert_eq!(
            provider.license_file().unwrap(),
            Some("LICENSE.txt".to_string())
        );
    }

    #[test]
    fn test_missing_project_section() {
        let pyproject_toml_content = r#"
[build-system]
requires = ["setuptools", "wheel"]
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        assert_eq!(provider.name().unwrap(), None);
        assert_eq!(provider.version().unwrap(), None);
        assert_eq!(provider.description().unwrap(), None);
    }

    #[test]
    fn test_input_globs() {
        let pyproject_toml_content = r#"
    [project]
    name = "test-package"
    version = "1.0.0"
    "#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        // Force loading of manifest
        let _ = provider.name().unwrap();

        let globs = provider.input_globs();
        assert_eq!(globs.len(), 1);
        assert!(globs.contains("pyproject.toml"));
    }

    #[test]
    fn test_ignore_pyproject_manifest_flag() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "1.0.0"
description = "Test description"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = PyprojectMetadataProvider::new(temp_dir.path(), true);

        // All methods should return None when ignore_pyproject_manifest is true
        assert_eq!(provider.name().unwrap(), None);
        assert_eq!(provider.version().unwrap(), None);
        assert_eq!(provider.description().unwrap(), None);
        assert_eq!(provider.license().unwrap(), None);
        assert_eq!(provider.homepage().unwrap(), None);
        assert_eq!(provider.repository().unwrap(), None);
        assert_eq!(provider.documentation().unwrap(), None);
        assert_eq!(provider.license_file().unwrap(), None);
        assert_eq!(provider.summary().unwrap(), None);
    }

    #[test]
    fn test_alternative_url_keys() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "1.0.0"

[project.urls]
"Source Code" = "https://github.com/example/test-package"
Docs = "https://docs.example.com"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        assert_eq!(
            provider.repository().unwrap(),
            Some("https://github.com/example/test-package".to_string())
        );
        assert_eq!(
            provider.documentation().unwrap(),
            Some("https://docs.example.com".to_string())
        );
    }

    #[test]
    fn test_invalid_version_format() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "1.0.0a1"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        // This should parse successfully since it's a valid PEP440 version
        let result = provider.version();
        assert!(result.is_ok());
        assert!(result.unwrap().is_some());
    }

    #[test]
    fn test_pyproject_toml_parse_error() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "not.a.valid.version.at.all"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        let result = provider.version();
        // The pyproject-toml parser should fail to parse this
        match result {
            Err(MetadataError::PyProjectToml(_)) => {
                // This is expected - invalid version in pyproject.toml
            }
            other => panic!(
                "Expected PyProjectTomlError for invalid version, got: {:?}",
                other
            ),
        }
    }

    #[test]
    fn test_malformed_pyproject_toml() {
        let pyproject_toml_content = r#"
[project
name = "test-package"
version = "1.0.0"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        let result = provider.name();
        assert!(result.is_err());
        match result.unwrap_err() {
            MetadataError::PyProjectToml(_) => {}
            err => panic!("Expected PyProjectToml, got: {:?}", err),
        }
    }

    #[test]
    fn test_summary_equals_description() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "1.0.0"
description = "Test description"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        let provider = create_metadata_provider(temp_dir.path());

        let description = provider.description().unwrap();
        let summary = provider.summary().unwrap();

        assert_eq!(description, summary);
        assert_eq!(summary, Some("Test description".to_string()));
    }

    #[test]
    fn test_generated_recipe_contains_pyproject_values() {
        let pyproject_toml_content = r#"
[project]
name = "test-package"
version = "99.0.0"
description = "A test package"
license = {text = "MIT"}

[project.urls]
Homepage = "https://example.com"
Repository = "https://github.com/example/test-package"
Documentation = "https://docs.example.com"
"#;

        let temp_dir = create_temp_pyproject_project(pyproject_toml_content);
        // let mut provider = create_metadata_provider(temp_dir.path());

        // Now create project model and generate a recipe from it
        let project_model = project_fixture!({
            "name": "foobar",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let generated_recipe = PythonGenerator::default()
            .generate_recipe(
                &project_model,
                // when using the default here we should read values from the pyproject.toml
                &PythonBackendConfig::default(),
                temp_dir.path().to_path_buf(),
                Platform::Linux64,
                None,
                &HashSet::new(),
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe, {
        ".source[0].path" => "[ ... path ... ]",
        ".build.script" => "[ ... script ... ]",
        });
    }
}
