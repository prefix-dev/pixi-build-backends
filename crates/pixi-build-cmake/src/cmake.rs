use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
    str::FromStr,
};

use crate::{
    build_script::{BuildPlatform, BuildScriptContext},
    config::CMakeBackendConfig,
};
use miette::IntoDiagnostic;
use pixi_build_backend::common::{PackageRequirements, SourceRequirements};
use pixi_build_backend::{ProjectModel, Targets};
use pixi_build_backend::{
    common::{BuildConfigurationParams, requirements},
    compilers::default_compiler,
    traits::{Dependencies, project::new_spec},
};
use rattler_build::metadata::Debug;
use rattler_build::recipe::parser::{BuildString, ScriptContent};
use rattler_build::{
    NormalizedKey,
    console_utils::LoggingOutputHandler,
    hash::HashInfo,
    metadata::{BuildConfiguration, PackagingSettings},
    recipe::{
        Recipe,
        parser::{Build, Dependency, Package, Script},
        variable::Variable,
    },
};
use rattler_conda_types::{MatchSpec, NoArchType, PackageName, Platform, package::ArchiveType};
use rattler_package_streaming::write::CompressionLevel;

pub struct CMakeBuildBackend<P: ProjectModel> {
    pub(crate) logging_output_handler: LoggingOutputHandler,
    pub(crate) manifest_path: PathBuf,
    pub(crate) manifest_root: PathBuf,
    pub(crate) project_model: P,
    pub(crate) config: CMakeBackendConfig,
    pub(crate) cache_dir: Option<PathBuf>,
}

impl<P: ProjectModel> CMakeBuildBackend<P> {
    /// Returns a new instance of [`CMakeBuildBackend`] by reading the manifest
    /// at the given path.
    pub fn new(
        manifest_path: &Path,
        project_model: P,
        config: CMakeBackendConfig,
        logging_output_handler: LoggingOutputHandler,
        cache_dir: Option<PathBuf>,
    ) -> miette::Result<Self> {
        // Determine the root directory of the manifest
        let manifest_root = manifest_path
            .parent()
            .ok_or_else(|| miette::miette!("the project manifest must reside in a directory"))?
            .to_path_buf();

        Ok(Self {
            manifest_path: manifest_path.to_path_buf(),
            manifest_root,
            project_model,
            config,
            logging_output_handler,
            cache_dir,
        })
    }

    /// Returns the matchspecs for the compiler packages. That should be
    /// included in the build section of the recipe.
    fn compiler_packages(&self, target_platform: Platform) -> Vec<MatchSpec> {
        let mut compilers = vec![];

        for lang in self.languages() {
            if let Some(name) = default_compiler(target_platform, &lang) {
                // TODO: Read this from variants
                // TODO: Read the version specification from variants
                let compiler_package =
                    PackageName::new_unchecked(format!("{name}_{target_platform}"));
                compilers.push(MatchSpec::from(compiler_package));
            }

            // TODO: stdlib??
        }

        compilers
    }

    /// Returns the languages that are used in the cmake project. These define
    /// which compilers are required to build the project.
    fn languages(&self) -> Vec<String> {
        // TODO: Can we figure this out from looking at the CMake?
        vec!["cxx".to_string()]
    }

    /// Constructs a [`Recipe`] from the current manifest. The constructed
    /// recipe will invoke CMake to build and install the package.
    pub(crate) fn recipe(
        &self,
        host_platform: Platform,
        variant: &BTreeMap<NormalizedKey, Variable>,
    ) -> miette::Result<(Recipe, SourceRequirements<P>)> {
        // Parse the package name from the manifest
        let project_model = &self.project_model;
        let name = PackageName::from_str(project_model.name()).into_diagnostic()?;
        let version = self.project_model.version().clone().ok_or_else(|| {
            miette::miette!("a version is missing from the package but it is required")
        })?;

        let noarch_type = NoArchType::none();

        let requirements = self.requirements(host_platform, variant)?;

        let build_platform = Platform::current();
        let build_number = 0;

        // Check if the host platform has a host python dependency
        // This is used to determine if we need to the cmake argument for the python executable
        let has_host_python = self
            .project_model
            .dependencies(Some(host_platform))
            .contains(&"python".into());

        let build_script = BuildScriptContext {
            build_platform: if build_platform.is_windows() {
                BuildPlatform::Windows
            } else {
                BuildPlatform::Unix
            },
            source_dir: self.manifest_root.display().to_string(),
            extra_args: self.config.extra_args.clone(),
            has_host_python,
        }
        .render();

        let hash_info = HashInfo::from_variant(variant, &noarch_type);

        Ok((
            Recipe {
                schema_version: 1,
                context: Default::default(),
                package: Package {
                    version: version.into(),
                    name,
                },
                cache: None,
                // source: vec![Source::Path(PathSource {
                //     // TODO: How can we use a git source?
                //     path: manifest_root.to_path_buf(),
                //     sha256: None,
                //     md5: None,
                //     patches: vec![],
                //     target_directory: None,
                //     file_name: None,
                //     use_gitignore: true,
                // })],
                // We hack the source location
                // as we dont want to copy the files
                source: vec![],
                build: Build {
                    number: build_number,
                    string: BuildString::Resolved(BuildString::compute(&hash_info, build_number)),

                    // skip: Default::default(),
                    script: Script {
                        content: ScriptContent::Commands(build_script),
                        env: self.config.env.clone(),
                        ..Default::default()
                    },
                    noarch: noarch_type,

                    // TODO: Python is not exposed properly
                    //python: Default::default(),
                    // dynamic_linking: Default::default(),
                    // always_copy_files: Default::default(),
                    // always_include_files: Default::default(),
                    // merge_build_and_host_envs: false,
                    // variant: Default::default(),
                    // prefix_detection: Default::default(),
                    // post_process: vec![],
                    // files: Default::default(),
                    ..Build::default()
                },
                // TODO read from manifest
                requirements: requirements.requirements,
                tests: vec![],
                about: Default::default(),
                extra: Default::default(),
            },
            requirements.source,
        ))
    }

    pub(crate) fn requirements(
        &self,
        host_platform: Platform,
        variant: &BTreeMap<NormalizedKey, Variable>,
    ) -> miette::Result<PackageRequirements<P>> {
        let project_model = &self.project_model;
        let dependencies = project_model.dependencies(Some(host_platform));

        let build_tools = build_tools();
        let empty_spec = new_spec::<P>();
        let dependencies = add_build_tools::<P>(dependencies, &build_tools, &empty_spec);

        let mut package_requirements = requirements::<P>(dependencies, variant)?;

        package_requirements.requirements.build.extend(
            self.compiler_packages(host_platform)
                .into_iter()
                .map(Dependency::Spec),
        );

        Ok(package_requirements)
    }
}

/// Returns the build tools that are required to build the project.
pub(crate) fn build_tools() -> Vec<String> {
    vec!["cmake".to_string(), "ninja".to_string()]
}

/// Adds the build tools to the dependencies.
pub(crate) fn add_build_tools<'a, P: ProjectModel>(
    mut dependencies: Dependencies<'a, <P::Targets as Targets>::Spec>,
    build_tools: &'a [String],
    empty_spec: &'a <<P as ProjectModel>::Targets as Targets>::Spec,
) -> Dependencies<'a, <<P as ProjectModel>::Targets as Targets>::Spec> {
    for pkg_name in build_tools.iter() {
        if dependencies.build.contains_key(pkg_name) {
            // If the host dependencies already contain the package, we don't need to add it
            // again.
            continue;
        }

        dependencies.build.insert(pkg_name, empty_spec);
    }

    dependencies
}

pub(crate) fn construct_configuration(
    recipe: &Recipe,
    params: BuildConfigurationParams,
) -> BuildConfiguration {
    BuildConfiguration {
        target_platform: params.host_platform.platform,
        host_platform: params.host_platform,
        build_platform: params.build_platform,
        hash: HashInfo::from_variant(&params.variant, &recipe.build.noarch),
        variant: params.variant,
        directories: params.directories,
        channels: params.channels,
        channel_priority: Default::default(),
        solve_strategy: Default::default(),
        timestamp: chrono::Utc::now(),
        subpackages: Default::default(), // TODO: ???
        packaging_settings: PackagingSettings::from_args(
            ArchiveType::Conda,
            CompressionLevel::default(),
        ),
        store_recipe: false,
        force_colors: true,
        sandbox_config: None,
        debug: Debug::new(false),
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use indexmap::IndexMap;
    use pixi_build_type_conversions::to_project_model_v1;
    use pixi_manifest::Manifests;
    use rattler_build::{console_utils::LoggingOutputHandler, recipe::Recipe};
    use rattler_conda_types::{ChannelConfig, Platform};
    use tempfile::tempdir;

    use crate::{cmake::CMakeBuildBackend, config::CMakeBackendConfig};

    fn recipe(manifest_source: &str, config: CMakeBackendConfig) -> Recipe {
        let tmp_dir = tempdir().unwrap();
        let tmp_manifest = tmp_dir.path().join("pixi.toml");
        std::fs::write(&tmp_manifest, manifest_source).unwrap();
        let manifest = Manifests::from_workspace_manifest_path(tmp_manifest.clone()).unwrap();
        let package = manifest.value.package.unwrap();
        let channel_config = ChannelConfig::default_with_root_dir(tmp_dir.path().to_path_buf());
        let project_model = to_project_model_v1(&package.value, &channel_config).unwrap();

        let cmake_backend = CMakeBuildBackend::new(
            &tmp_manifest,
            project_model,
            config,
            LoggingOutputHandler::default(),
            None,
        )
        .unwrap();

        cmake_backend
            .recipe(Platform::current(), &BTreeMap::new())
            .unwrap()
            .0
    }

    #[tokio::test]
    async fn test_setting_host_and_build_requirements() {
        // get cargo manifest dir

        let package_with_host_and_build_deps = r#"
        [workspace]
        name = "test-reqs"
        channels = ["conda-forge"]
        platforms = ["osx-arm64"]
        preview = ["pixi-build"]

        [package]
        name = "test-reqs"
        version = "1.0"

        [package.host-dependencies]
        hatchling = "*"

        [package.build-dependencies]
        boltons = "*"

        [package.run-dependencies]
        foobar = "==3.2.1"

        [package.build]
        backend = { name = "pixi-build-python", version = "*" }
        "#;

        let tmp_dir = tempdir().unwrap();
        let tmp_manifest = tmp_dir.path().join("pixi.toml");

        // write the raw string into the file
        std::fs::write(&tmp_manifest, package_with_host_and_build_deps).unwrap();

        let manifest = Manifests::from_workspace_manifest_path(tmp_manifest).unwrap();
        let package = manifest.value.package.unwrap();
        let channel_config = ChannelConfig::default_with_root_dir(tmp_dir.path().to_path_buf());
        let project_model = to_project_model_v1(&package.value, &channel_config).unwrap();
        let cmake_backend = CMakeBuildBackend::new(
            &package.provenance.path,
            project_model,
            CMakeBackendConfig::default(),
            LoggingOutputHandler::default(),
            None,
        )
        .unwrap();

        let host_platform = Platform::current();

        let (recipe, _source_requirements) = cmake_backend
            .recipe(host_platform, &BTreeMap::new())
            .unwrap();
        insta::with_settings!({
            filters => vec![
                ("(vs2017|vs2019|gxx|clang).*", "\"[ ... compiler ... ]\""),
            ]
        }, {
            insta::assert_yaml_snapshot!(recipe, {
               ".build.script" => "[ ... script ... ]",
            });
        });
    }

    #[tokio::test]
    async fn test_parsing_subdirectory() {
        // a manifest with subdir

        let package_with_git_and_subdir = r#"
        [workspace]
        name = "test-reqs"
        channels = ["conda-forge"]
        platforms = ["osx-arm64"]
        preview = ["pixi-build"]

        [package]
        name = "test-reqs"
        version = "1.0"

        [package.build]
        backend = { name = "pixi-build-python", version = "*" }

        [package.host-dependencies]
        hatchling = { git = "git+https://github.com/hatchling/hatchling.git", subdirectory = "src" }
        "#;

        let tmp_dir = tempdir().unwrap();
        let tmp_manifest = tmp_dir.path().join("pixi.toml");

        // write the raw string into the file
        std::fs::write(&tmp_manifest, package_with_git_and_subdir).unwrap();

        Manifests::from_workspace_manifest_path(tmp_manifest).unwrap();
    }

    #[test]
    fn test_env_vars_are_set() {
        let manifest_source = r#"
        [workspace]
        platforms = []
        channels = []
        preview = ["pixi-build"]

        [package]
        name = "foobar"
        version = "0.1.0"

        [package.build]
        backend = { name = "pixi-build-rust", version = "*" }
        "#;

        let env = IndexMap::from([("foo".to_string(), "bar".to_string())]);

        let recipe = recipe(
            manifest_source,
            CMakeBackendConfig {
                env: env.clone(),
                ..Default::default()
            },
        );

        assert_eq!(recipe.build.script.env, env);
    }

    #[test]
    fn test_python_host_dependency() {
        let manifest_source = r#"
        [workspace]
        platforms = []
        channels = []
        preview = ["pixi-build"]

        [package]
        name = "foobar"
        version = "0.1.0"

        [package.host-dependencies]
        python = "*"

        [package.build]
        backend = { name = "pixi-build-cmake", version = "*" }
        "#;

        let recipe = recipe(
            manifest_source,
            CMakeBackendConfig {
                ..Default::default()
            },
        );

        let command = match recipe.build.script.content {
            rattler_build::recipe::parser::ScriptContent::Commands(items) => items,
            _ => unreachable!("Expected script content to be commands"),
        };

        assert!(command.iter().any(|s| s.contains("-DPython_EXECUTABLE")));
    }
}
