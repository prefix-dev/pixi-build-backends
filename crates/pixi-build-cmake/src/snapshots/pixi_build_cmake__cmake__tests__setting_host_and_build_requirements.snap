---
source: crates/pixi-build-cmake/src/cmake.rs
expression: recipe
---
schema_version: 1
context: {}
package:
  name: test-reqs
  version: "1.0"
build:
  number: 0
  string: hbf21a9e_0
  script: "[ ... script ... ]"
requirements:
  build:
    - boltons
    - cmake
    - ninja
    - "[ ... compiler ... ]"
  host:
    - hatchling
  run:
    - foobar ==3.2.1
